### DISCUSSION TITLE
Backup status / last backup

### DISCUSSION THREAD

**Discussion by tubby1981** [atc0005/check-vmware#504] [Stable Release]
At the overview of a VM you have a heading "Custom Attributes" with 2 fields. Namely "Backup status" and "Last backup" In these fields the backup software stores the time when the last successful backup was and information about the backup such as backup job etc. Is it possible to make a nagios check of this as well?

* Warning: If these fields are empty, the VM is not yet included in the backup software
* Critical: if the VM has not been backed up for X number of days
* also an option to exclude vms

**Comment by atc0005** [comment]
@tubby1981 This seems like it would be a useful plugin.

Later on with some refactoring it could probably be generalized to evaluate Custom Attributes using some other criteria.

Do you have an example of the date format used in the field? 

Are you comfortable sharing the contents of the fields, a screenshot, etc.? Not a problem if you'd rather not.

**Comment by tubby1981** [comment]
See the images of a VM that is not yet in the backup or where a backup has not yet been made. And a VM where a backup has been made.

![vm_last_backup](https://user-images.githubusercontent.com/5524630/140895050-f1d7bc72-c757-4f6b-b075-9fe02160baee.png)
![vm_not_in_backup](https://user-images.githubusercontent.com/5524630/140895053-f06cf6f0-4317-4a8f-8cd6-04b3a067dcec.png)

**Reply by tubby1981** [reply]
If I need to test something, just let me know. I think the exclude list should support regex so that we can exclude multiple vms with a regex.

**Reply by atc0005** [reply]
> @tubby1981: If I need to test something, just let me know.

Sure, will do. Currently I'm focused on #505. Once I'm done with that plugin I hope to look at this, but I can't make any promises on timeline (limited dev time these days).

>  I think the exclude list should support regex so that we can exclude multiple vms with a regex.

This sounds useful. The first implementation will likely borrow from existing helper functions to include/exclude based solely on full VM name, though adding support for regex patterns at a later point sounds like a very useful addition.

**Comment by atc0005** [comment]
@tubby1981 

This is the next larger item in my queue.

As I've mentioned previously, full support will likely be added in phases with the initial version being a minimal implementation.

The first test version of the plugin will likely only gather details and emit them to the console. As you're able to test and provide that output it should help with further development.

No estimates on completion or when I'll have something to test, just letting you know that I've not forgotten about the request.

**Comment by tubby1981** [comment]
Excellent. Let me know when I can do some testing

**Reply by atc0005** [reply]
Will do. I've begun experimenting with the `vcsim` tool and it's already helped with local testing/development. I'm able to define the needed Custom Attributes within that environment with values that match that pattern you've shared and a prototype build is able to successfully retrieve the values. This should reduce the number of times I will need you to run a plugin on your end to determine if I'm moving in the right direction.

I'd estimate a few weeks before a plugin is available to test, but I'll let you know.

**Comment by atc0005** [comment]
@tubby1981 Are you aware of any cases where one of the attributes is set, but the other is not? For example, the `Last Backup` Custom Attribute has a valid value, but the `Backup Status` field is empty.

I'm going to assume that if one field is set then the other is also set, but wanted to double-check and be (at least relatively) sure.

When you installed the commvault software:

- did you have to specify which Custom Attributes to use?
- did you have to create the Custom Attributes yourself in your vCenter instance, or did the installer(s) take care of that?

**Reply by tubby1981** [reply]
No, the "Custom Attributes" are filled by default from the commvault software. We have not configured this ourselves.

**Comment by atc0005** [comment]
@tubby1981 Additional questions:

- what time zone is used for the `Last Backup` Custom Attribute?
- is the local time zone assumed?

I'm trying to get a sense of whether the time zone used in date calculations should default to local time with an option to override with a specific time zone value. Need to account for the scenario where the backup date/time value is written in one assumed time zone and the monitoring system is working with another.

**Reply by tubby1981** [reply]
The time zone is Europe/Amsterdam (our local timezone). 

Timeformat = 01/11/2022 04:03:44 (month/day/year hours:minutes:seconds)

**Comment by atc0005** [comment]
@tubby1981 I have an initial version of the plugin that I'm working on prepping for an alpha release. If I have time I'll upload the alpha release today, but it could be tomorrow before I can do so.

No documentation is available yet, but here is a rough idea of how to use it:

```shell
/usr/lib/nagios/plugins/check_vmware_vm_backup_via_ca 
    --username ${VCSIM_USER} \
    --password ${VCSIM_PASS} \
    --server ${VCSIM_SERVER} \
    --port ${VCSIM_PORT} \
    --trust-cert \
    --log-level info \
    --backup-date-timezone "Europe/Amsterdam" \
    --backup-metadata-ca "Backup Status"
```

The last flag is entirely optional and can be omitted to trim some of the
output. Specifying the time zone may be optional, depending on how the
monitoring system and backup system are configured.

As we talked about, excluding/ignoring VMs is possible, but for the time being
only via explicit names and not regular expressions. You can also limit
evaluation of VMs to specific resource pools, so that might be useful
depending on your environment.

**Reply by atc0005** [reply]
> @tubby1981 I have an initial version of the plugin that I'm working on prepping for an alpha release. If I have time I'll upload the alpha release today

`v0.29.0-alpha.1` is available here:

https://github.com/atc0005/check-vmware/releases/tag/v0.29.0-alpha.1

Please test and provide feedback.

**Reply by atc0005** [reply]
Looking over the results from some test cases, I see that the error chosen as the most prominent differs from the one notes in the one-line summary output (subject of alert).

Between the two, the main plugin logic is considering VMs with old backups a higher priority than VMs with missing backups (presumably not yet enrolled in a backup schedule). The one-line summary however is (unintentionally) considering them with reversed priority, noting first if missing backups (and how many) are detected.

I'm going to generate a fresh alpha release (just the one plugin) that uses updated/consistent logic.

**Reply by atc0005** [reply]
> I'm going to generate a fresh alpha release (just the one plugin) that uses updated/consistent logic.

@tubby1981 

Available here:

https://github.com/atc0005/check-vmware/releases/tag/v0.29.0-alpha.2

Contrast with the previous alpha to see the differences.

**Comment by tubby1981** [comment]
`
./check_vmware_vm_backup -u XXXXXXXXX -pw XXXXXXXXX -s  XXXXXXXXX --backup-date-timezone "Europe/Amsterdam" 
{"level":"error","version":"check-vmware v0.29.0-alpha.2-0-g2e2bff3 (https://github.com/atc0005/check-vmware)","logging_level":"info","plugin_type":"vm-last-backup-via-ca","plugin_timeout":"10s","username":"XXXXXXXXX","user_domain":"","trust_cert":false,"server":"XXXXXXXXX","port":443,"included_resource_pools":"","excluded_resource_pools":"","ignored_vms":"","backup_age_critical":2,"backup_age_warning":1,"eval_powered_off":false,"ignore_missing_ca_on_objects":false,"error":"error evaluating backup date for virtual machine \"XXXXXXXXX\": parsing time \"01/17/2022 20:14:12\" as \"2006-01-02 3:04:05 PM\": cannot parse \"7/2022 20:14:12\" as \"2006\"","time":"2022-01-17T20:19:28+01:00","caller":"github.com/atc0005/check-vmware/cmd/check_vmware_vm_backup_via_ca/main.go:285","message":"error retrieving virtual machines with requested backup Custom Attributes"}
CRITICAL: Error retrieving virtual machines with requested backup Custom Attributes 
 
**ERRORS** 
 
* error evaluating backup date for virtual machine "XXXXXXXXX": parsing time "01/17/2022 20:14:12" as "2006-01-02 3:04:05 PM": cannot parse "7/2022 20:14:12" as "2006" 
 | 'time'=530ms;;;; 
`

How should I use the option: -backup-metadata-ca? Do you have information about that?

**Reply by atc0005** [reply]
The error appears to be from the time format given to the plugin not matching the time format found in your VMware environment. The Go `time` package parses an input date string using a given format or "layout" string. That layout string is expected to match the same format of the input string, using very specific numbers for year (2006), month, day, hour and so forth.

The default layout used by the plugin is `2006-01-02 3:04:05 PM`. Based on what you posted, it appears that you will need to override the default with a different layout string to match the input date string of `01/17/2022 20:14:12`.

I don't have easy access to a PC at the moment, but I can respond back tomorrow with a suggested layout string to use as the `--backup-date-format` flag value.

> How should I use the option: -backup-metadata-ca? 

This flag is intended for specifying the name of the custom attribute used by the backup software (commvault in your case)  to record additional metadata about the last backup. I think it is named `Backup Status`, but I don't recall from memory (typing this from a phone and it isn't easy to scroll back on this app).

> Do you have information about that?

Check the `-h` or `--help` output. Let me know if it makes sense (it could use a second set of eyes to make sure I didn'tiss anything).

**Reply by atc0005** [reply]
> The default layout used by the plugin is 2006-01-02 3:04:05 PM. Based on what you posted, it appears that you will need to override the default with a different layout string to match the input date string of 01/17/2022 20:14:12.

This is technically true, but looking back over the example date format and other details I see that the default should have instead matched your environment without requiring that you override the default.

I will plan to release a new build that fixes this and some other items I spotted.

**Reply by atc0005** [reply]
> > The default layout used by the plugin is 2006-01-02 3:04:05 PM. Based on what you posted, it appears that you will need to override the default with a different layout string to match the input date string of 01/17/2022 20:14:12.
> 
> This is technically true, but looking back over the example date format and other details I see that the default should have instead matched your environment without requiring that you override the default.
> 
> I will plan to release a new build that fixes this and some other items I spotted.

Follow-up (and mostly a "thinking out loud" response):

I see now why the default layout doesn't match the one provided. 

The screenshot provided on https://github.com/atc0005/check-vmware/discussions/504#discussioncomment-1611102 clearly indicates a date/time format that matches the `01/17/2022 20:14:12` value, as does the later sample provided on https://github.com/atc0005/check-vmware/discussions/504#discussioncomment-1950251. 

Without meaning to, I made the assumption that the format shown in the provided screenshot from the https://community.commvault.com/technical-q-a-2/vmware-backup-job-status-output-via-custom-attributes-1796 link was a universal format used by commvault and didn't pay close enough attention between the previously provided screenshot and the one in the commvault thread as the plugin offered the ability to override the format as needed.

As noted previously, I'll update the default layout string to match the `01/17/2022 20:14:12` value. I will also make plans to provide a table in the documentation that maps common date formats with equivalent layout strings for the `--backup-date-format` flag.

**Comment by tubby1981** [comment]
--backup-date-format "DD/MM/YYYY HH:MM:SS"

`**ERRORS** 
 
* error evaluating backup date for virtual machine "XXXXXX": parsing time "01/17/2022 20:14:12" as "DD/MM/YYYY HH:MM:SS": cannot parse "01/17/2022 20:14:12" as "DD/MM/YYYY HH:MM:SS" 
 | 'time'=575ms;;;;`

**Reply by atc0005** [reply]
@tubby1981 

The Go `time` package uses a different approach for specifying the "layout" for parsing dates/times:

https://pkg.go.dev/time#pkg-constants

Instead of "DD", "MM" or "YY" or "YYYY"), it uses literal numbers to in the layout string to match the expected format of input date/time values.

For example, to match the date/time value of `01/17/2022 20:14:12`, I believe that you would use this layout string:

  `01/02/2006 15:04:05`

Each of those numbers is a human readable placeholder for the specific value that they're used for matching:

- `01` for the month (which just happens to be the month we're in)
- `02` for the day
- `2006` for the year
- `15` for the two digit hour
- `04` for the two digit minute
- `05` for the two digit second

I will reset the test environment I'm using and attempt to use the `01/02/2006 16:04:05` value I mentioned above to match the expected date/time string you shared.

**EDIT**:

Replaced `16` with `15` in the suggested layout string noted above (`16` is incorrect).

**Reply by atc0005** [reply]
> @tubby1981:
>
> --backup-date-format "DD/MM/YYYY HH:MM:SS"

I just finished testing on my end. Use this:

```console
/usr/lib/nagios/plugins/check_vmware_vm_backup_via_ca \
    --username "${VCSIM_USER}" \
    --password "${VCSIM_PASS}" \
    --server "${VCSIM_SERVER}" \
    --port "${VCSIM_PORT}" \
    --trust-cert \
    --log-level info \
    --backup-date-timezone "Europe/Amsterdam" \
    --backup-metadata-ca "Backup Status" \
    --backup-date-format "01/02/2006 15:04:05"
```

Substitute with applicable values as you did before. E.g., the `--port` flag is entirely optional if you're using the default, the `--log-level` flag might instead be given as `--log-level debug` or omitted entirely (the default log level is already `info`) and so forth.

**Reply by atc0005** [reply]
> > @tubby1981:
> > --backup-date-format "DD/MM/YYYY HH:MM:SS"
> 
> I just finished testing on my end. Use this:
> 
> ```
> /usr/lib/nagios/plugins/check_vmware_vm_backup_via_ca \
>     --username "${VCSIM_USER}" \
>     --password "${VCSIM_PASS}" \
>     --server "${VCSIM_SERVER}" \
>     --port "${VCSIM_PORT}" \
>     --trust-cert \
>     --log-level info \
>     --backup-date-timezone "Europe/Amsterdam" \
>     --backup-metadata-ca "Backup Status" \
>     --backup-date-format "01/02/2006 15:04:05"
> ```
> 
> Substitute with applicable values as you did before. E.g., the `--port` flag is entirely optional if you're using the default, the `--log-level` flag might instead be given as `--log-level debug` or omitted entirely (the default log level is already `info`) and so forth.

New alpha release:

https://github.com/atc0005/check-vmware/releases/tag/v0.29.0-alpha.3

Aside from some minor output changes, if you use the `--backup-date-format "01/02/2006 15:04:05"` flag & value as noted previously, there should no differences between the `v0.29.0-alpha.2` and `v0.29.0-alpha.3` releases.

If you don't use the `--backup-date-format` flag with the `v0.29.0-alpha.3` release, the results should be the same as if you used the `--backup-date-format "01/02/2006 15:04:05"` flag & value with the `v0.29.0-alpha.2` release.

**Comment by tubby1981** [comment]
`WARNING: 9 VMs missing backups detected (91 present & 91 current, evaluated 100 VMs & 1 Resource Pools) 
 
**ERRORS** 
 
* virtual machine missing backup date 
 
**THRESHOLDS** 
 
* CRITICAL: non-excluded VM with:  
        backup date exceeding specified CRITICAL threshold 
* WARNING: non-excluded VM with:  
        backup date exceeding specified WARNING threshold, but not CRITICAL threshold 
        backup date missing 
        backup date does not match default/user-specified format 
 
**DETAILED INFO** 
 
VMs without backups: 
 
* XXXXXXXXXXXXXXXX - clone for upgrade 
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX 
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX 
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX 
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* vXXXXXXXXXXXXXXXX 
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
* XXXXXXXXXXXXXXXX
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
 
VMs with old backups:  
 
* None 
 
Virtual Machines Backup Summary:  
 
* Missing Backups: 9 
* With Backups: 91 
* Old Backups: 0 
* Oldest backup: XXXXXXXXXXXXXXXX (19h ago) 
* Most recent backup: XXXXXXXXXXXXXXXX (11h ago) 
 
--- 
 
* vSphere environment: https://XXXXXXXXXXXXXXXX:443/sdk 
* Plugin User Agent: check-vmware/v0.29.0-alpha.3-0-gf95bd6d 
* VMs (total: 100, evaluated: 100) 
* Powered off VMs evaluated: true 
* Specified VMs to exclude (0): [] 
* Specified Resource Pools to explicitly include (0): [] 
* Specified Resource Pools to explicitly exclude (0): [] 
* Resource Pools evaluated (1): [Resources] 
 
 | 'vms'=100;;;; 'vms_excluded_by_name'=0;;;; 'vms_evaluated'=100;;;; 'vms_with_backup_dates'=91;;;; 'vms_without_backup_dates'=9;;;; 'resource_pools_excluded'=0;;;; 'resource_pools_included'=0;;;; 'resource_pools_evaluated'=1;;;; 'time'=506ms;;;; `

Looks good. Excluding VMs also works well. Is it possible not to warn for VMs that are new? Who are not yet 24 hours old?

**Reply by atc0005** [reply]
> Looks good. Excluding VMs also works well. 

Thanks for testing and for the detailed feedback/output that you've shared.

I was on the fence (and admittedly still am) regarding the list of VMs. There is this format that the alpha builds are currently using:

```
* XXXXXXXXXXXXXXXX
** Old Backup: false
** Backup age (formatted): 106751d 23h ago
** Backup age (in days): 106751
** Last Backup (raw value): ""
```

and a more condensed version that I was considering:

```
* WWWWWWWWWWWWWWWWW [Old backup: false, Backup age: 5, Raw Value: "01/17/2022 20:14:12"]
* XXXXXXXXXXXXXXXXX [Old backup: false, Backup age: 5, Raw Value: "01/17/2022 20:14:12"]
* YYYYYYYYYYYYYYYYY [Old backup: false, Backup age: 5, Raw Value: "01/17/2022 20:14:12"]
* ZZZZZZZZZZZZZZZZZ [Old backup: false, Backup age: 5, Raw Value: "01/17/2022 20:14:12"]
```

The current format seems fine for ~ 10-15 VMs, but past that it might prove challenging to read at a glance. The condensed version could probably list 50 (just a guess) or so in the same screen space.

What do you think about the condensed version? The downside is that this would result in the metadata field not being emitted. I suppose both formats could be supported with the trigger to switch between them being whether the backup metadata flag was specified; if the flag is provided, the more verbose form is used, otherwise the condensed version, which would become the default.

> Is it possible not to warn for VMs that are new? Who are not yet 24 hours old?

Potentially, but in an effort to get the current functionality stabilized I'd suggest that we treat that as a separate request similar to:

- GH-588
 
Adding support for excluding VMs based on age could also be useful for all existing plugins that currently support ignoring virtual machines based on given name. For environments with a lot of VM creation activity, this could be a big help, particularly where automation is being used to spin off temporary instances (that might never need a backup).

**Reply by atc0005** [reply]
> Potentially, but in an effort to get the current functionality stabilized I'd suggest that we treat that as a separate request

Spun off a separate GH issue for later work:

- GH-595

**Reply by atc0005** [reply]
First batch of documentation updates complete.

Going to tag the current state as a new alpha release (`v0.29.0-alpha.4`).

For what I hope will be the final alpha release, I'll make further tweaks to the listing of VMs to better handle scenarios like this one:

```
* XXXXXXXXXXXXXXXX
** Old Backup: false
** Backup age (formatted): 106751d 23h ago
** Backup age (in days): 106751
** Last Backup (raw value): ""
```

Here there is no recorded backup, so providing a backup age (formatted or number of days) isn't really valuable. Instead, we should probably just note that there isn't a backup and emit the "raw value" to explain why we're noting as much.

**Reply by atc0005** [reply]
> For what I hope will be the final alpha release, I'll make further tweaks to the listing of VMs to better handle scenarios like this one:
> 
> ```
> * XXXXXXXXXXXXXXXX
> ** Old Backup: false
> ** Backup age (formatted): 106751d 23h ago
> ** Backup age (in days): 106751
> ** Last Backup (raw value): ""
> ```
> 
> Here there is no recorded backup, so providing a backup age (formatted or number of days) isn't really valuable. Instead, we should probably just note that there isn't a backup and emit the "raw value" to explain why we're noting as much.

@tubby1981 tagged these changes as `v0.29.0-alpha.5`. Please test this build and see if the output is improved from the prior build.

**Comment by atc0005** [comment]
For the next alpha, I'm planning on updating warning/critical "backup date" references with warning/critical "backup age" instead; we're specifying number of days (old) and not a specific date for thresholds. This change better matches flag names used by other plugins (e.g., snapshots age).

**EDIT**:  

- See GH-597
- Regarding "For the next alpha", I was referring to `v0.29.0-alpha.4` which has since been released.

**Comment by tubby1981** [comment]
`* XXXXXXXX
        ** Old Backup: false 
        ** Backup age (formatted): 106751d 23h ago 
        ** Backup age (in days): 106751 
        ** Last Backup (raw value): "" 
 
VMs with old backups:  
 
* None 
 
Virtual Machines Backup Summary:  
 
* Missing Backups: 6 
* With Backups: 94 
* Old Backups: 0 
* Oldest backup: XXXXXXXX (18h ago) 
* Most recent backup: XXXXXXXX (10h ago) 
 
--- 
 
* vSphere environment: https://XXXXXXXX:443/sdk 
* Plugin User Agent: check-vmware/v0.29.0-alpha.4-0-g6dc90df 
* VMs (total: 100, evaluated: 100) 
* Powered off VMs evaluated: true 
* Specified VMs to exclude (0): [] 
* Specified Resource Pools to explicitly include (0): [] 
* Specified Resource Pools to explicitly exclude (1): [NOC] 
* Resource Pools evaluated (21): [XXXXXXXX,XXXXXXXX,XXXXXXXX,XXXXXXXXXXXXXXXXXXXXXXXX] 
 
 | 'vms'=100;;;; 'vms_excluded_by_name'=0;;;; 'vms_evaluated'=100;;;; 'vms_with_backup_dates'=94;;;; 'vms_without_backup_dates'=6;;;; 'resource_pools_excluded'=1;;;; 'resource_pools_included'=0;;;; 'resource_pools_evaluated'=21;;;; 'time'=1213ms;;;; `

exclude-rp doesn't seem to work. I still get notifications about VMs not being backed up when I use this option

**Reply by atc0005** [reply]
> @tubby1981: exclude-rp doesn't seem to work. I still get notifications about VMs not being backed up when I use this option

Just to confirm, is the following true?

- you are excluding a specific resource pool named `NOC`
- VMs within the `NOC` resource pool are showing up in the list as not having backups

**Reply by tubby1981** [reply]
Yes thats correct.

**Reply by atc0005** [reply]
Do you happen to use the `exclude-rp` flag with any other plugins from this project? Are those plugins also evaluating VMs from resource pools that should be excluded?

I'm trying to get a sense of whether refactoring work from GH-494 might be to blame.

If you're comfortable doing so, sharing your debug log output would be a big help as would sharing what CLI flags and values were used. Sensitive values (e.g., `username`, `password`, maybe `domain`) are not needed.

**Reply by atc0005** [reply]
@tubby1981 I've been able to replicate the results, so no need for debug log output at this time. Will test further on my end (likely tomorrow) and respond further then.

**Reply by atc0005** [reply]
> @tubby1981 I've been able to replicate the results, so no need for debug log output at this time. Will test further on my end (likely tomorrow) and respond further then.

Opened GH-600 to track the issue. I'll plan to release a fix for that in the same release as the initial implementation of this plugin. I expect the all plugins which support that flag are impacted by the bug.

**Reply by atc0005** [reply]
> > @tubby1981 I've been able to replicate the results, so no need for debug log output at this time. Will test further on my end (likely tomorrow) and respond further then.
> 
> Opened [GH-600](https://github.com/atc0005/check-vmware/issues/600) to track the issue. I'll plan to release a fix for that in the same release as the initial implementation of this plugin. I expect the all plugins which support that flag are impacted by the bug.

I believed that this is fixed. Prepping `v0.29.0-alpha.6` with a rebuild of all plugins now, should have it available soon. Please test as you have time and indicate whether resource pool exclusion logic is working for you now. Thus far all tests in local vcsim environments are showing promising results.

**Comment by atc0005** [comment]
@tubby1981 All tasks that I had planned to include for the v0.29.0 release have been completed, including the initial implementation of this plugin. 

Please test and provide feedback for the first release candidate:

https://github.com/atc0005/check-vmware/releases/tag/v0.29.0-rc.1

Thanks.

**Reply by tubby1981** [reply]
The tests look good. Also the option --exclude-rp works now. I can't find any problems at the moment

**Reply by atc0005** [reply]
@tubby1981 Thanks for testing & for your feedback.

If nothing further is reported, I'll plan to generate a stable release in the next few days.

**Comment by atc0005** [comment]
Prepping the v0.29.0 release now. Considering the initial implementation of the plugin complete.
